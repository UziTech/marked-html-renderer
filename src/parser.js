export class Parser {
  constructor(options) {
    this.options = options;
    this.renderer = this.options.renderer;
    this.renderer.parser = this;
  }

  /**
   * Static Parse Method
   */
  static parse(tokens, options) {
    const parser = new Parser(options);
    return parser.parse(tokens);
  }

  /**
   * Static Parse Inline Method
   */
  static parseInline(tokens, options) {
    const parser = new Parser(options);
    return parser.parseInline(tokens);
  }

  appendOutput(out, ret) {
    if (ret) {
      out.append(ret);
    }
  }

  parse(tokens, top = true) {
    const out = document.createDocumentFragment();

    for (let i = 0; i < tokens.length; i++) {
      const anyToken = tokens[i];

      // Run any renderer extensions
      if (this.options.extensions?.renderers?.[anyToken.type]) {
        const genericToken = anyToken;
        const ret = this.options.extensions.renderers[genericToken.type].call({ parser: this }, genericToken);
        if (ret !== false || !['space', 'hr', 'heading', 'code', 'table', 'blockquote', 'list', 'html', 'paragraph', 'text'].includes(genericToken.type)) {
          this.appendOutput(out, ret);
          continue;
        }
      }

      const token = anyToken;

      switch (token.type) {
        case 'space': {
          this.appendOutput(out, this.renderer.space(token));
          continue;
        }
        case 'hr': {
          this.appendOutput(out, this.renderer.hr(token));
          continue;
        }
        case 'heading': {
          this.appendOutput(out, this.renderer.heading(token));
          continue;
        }
        case 'code': {
          this.appendOutput(out, this.renderer.code(token));
          continue;
        }
        case 'table': {
          this.appendOutput(out, this.renderer.table(token));
          continue;
        }
        case 'blockquote': {
          this.appendOutput(out, this.renderer.blockquote(token));
          continue;
        }
        case 'list': {
          this.appendOutput(out, this.renderer.list(token));
          continue;
        }
        case 'html': {
          this.appendOutput(out, this.renderer.html(token));
          continue;
        }
        case 'paragraph': {
          this.appendOutput(out, this.renderer.paragraph(token));
          continue;
        }
        case 'text': {
          let textToken = token;
          const body = document.createDocumentFragment();

          body.append(this.renderer.text(textToken));
          while (i + 1 < tokens.length && tokens[i + 1].type === 'text') {
            textToken = tokens[++i];
            body.append('\n', this.renderer.text(textToken));
          }
          if (top) {
            const p = this.renderer.paragraph({
              type: 'paragraph',
              raw: '',
              text: '',
              tokens: [],
            });
            p.append(body);
            this.appendOutput(out, p);
          } else {
            this.appendOutput(out, body);
          }
          continue;
        }

        default: {
          const errMsg = 'Token with "' + token.type + '" type was not found.';
          if (this.options.silent) {
            console.error(errMsg);
            return document.createDocumentFragment();
          } else {
            throw new Error(errMsg);
          }
        }
      }
    }

    return out;
  }

  /**
   * Parse Inline Tokens
   */
  parseInline(tokens, renderer = this.renderer) {
    const out = document.createDocumentFragment();

    for (let i = 0; i < tokens.length; i++) {
      const anyToken = tokens[i];

      // Run any renderer extensions
      if (this.options.extensions?.renderers?.[anyToken.type]) {
        const ret = this.options.extensions.renderers[anyToken.type].call({ parser: this }, anyToken);
        if (ret !== false || !['escape', 'html', 'link', 'image', 'strong', 'em', 'codespan', 'br', 'del', 'text'].includes(anyToken.type)) {
          this.appendOutput(out, ret);
          continue;
        }
      }

      const token = anyToken;

      switch (token.type) {
        case 'escape': {
          this.appendOutput(out, renderer.text(token));
          break;
        }
        case 'html': {
          this.appendOutput(out, renderer.html(token));
          break;
        }
        case 'link': {
          this.appendOutput(out, renderer.link(token));
          break;
        }
        case 'image': {
          this.appendOutput(out, renderer.image(token));
          break;
        }
        case 'strong': {
          this.appendOutput(out, renderer.strong(token));
          break;
        }
        case 'em': {
          this.appendOutput(out, renderer.em(token));
          break;
        }
        case 'codespan': {
          this.appendOutput(out, renderer.codespan(token));
          break;
        }
        case 'br': {
          this.appendOutput(out, renderer.br(token));
          break;
        }
        case 'del': {
          this.appendOutput(out, renderer.del(token));
          break;
        }
        case 'text': {
          this.appendOutput(out, renderer.text(token));
          break;
        }
        default: {
          const errMsg = 'Token with "' + token.type + '" type was not found.';
          if (this.options.silent) {
            console.error(errMsg);
            return document.createDocumentFragment();
          } else {
            throw new Error(errMsg);
          }
        }
      }
    }
    return out;
  }
}
