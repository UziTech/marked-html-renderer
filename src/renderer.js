import { textRenderer } from './textRenderer.js';

export const other = {
  escapeTest: /[&<>"']/,
  escapeReplace: /[&<>"']/g,
  escapeTestNoEncode: /[<>"']|&(?!(#\d{1,7}|#[Xx][a-fA-F0-9]{1,6}|\w+);)/,
  escapeReplaceNoEncode: /[<>"']|&(?!(#\d{1,7}|#[Xx][a-fA-F0-9]{1,6}|\w+);)/g,
  percentDecode: /%25/g,
  notSpaceStart: /^\S*/,
  endingNewline: /\n$/,
};

const escapeReplacements = {
  '&': '&amp;',
  '<': '&lt;',
  '>': '&gt;',
  '"': '&quot;',
  "'": '&#39;',
};
const getEscapeReplacement = (ch) => escapeReplacements[ch];

export function escapeText(html, encode) {
  if (encode) {
    if (other.escapeTest.test(html)) {
      return html.replace(other.escapeReplace, getEscapeReplacement);
    }
  } else {
    /* istanbul ignore else */
    if (other.escapeTestNoEncode.test(html)) {
      return html.replace(other.escapeReplaceNoEncode, getEscapeReplacement);
    }
  }

  return html;
}

export function cleanUrl(href) {
  try {
    href = encodeURI(href).replace(other.percentDecode, '%');
  } catch {
    /* istanbul ignore next */
    return null;
  }
  return href;
}

export const renderer = {

  space() {},

  code({ text, lang, escaped }) {
    const langString = (lang || '').match(other.notSpaceStart)?.[0];

    let code = text.replace(other.endingNewline, '') + '\n';
    code = (escaped ? code : escapeText(code, true));

    const preEl = document.createElement('pre');
    const codeEl = document.createElement('code');
    preEl.appendChild(codeEl);
    codeEl.textContent = code;

    if (langString) {
      preEl.classList.add('language-' + langString);
    }

    return preEl;
  },

  blockquote({ tokens }) {
    const blockquote = document.createElement('blockquote');
    blockquote.append(this.parser.parse(tokens));
    return blockquote;
  },

  html({ text }) {
    // HTML should be handled by the blockHtml and inlineHtml extensions in extension.js
    // Handle comments
    const comment = /^<!--([\s\S]*?)-->/.exec(text);
    /* istanbul ignore else */
    if (comment) {
      return document.createComment(comment[1]);
    }
    // If it is not just assume it is text.
    /* istanbul ignore next */
    return text;
  },

  heading({ tokens, depth }) {
    const heading = document.createElement('h' + depth);
    heading.append(this.parser.parseInline(tokens));
    return heading;
  },

  hr() {
    return document.createElement('hr');
  },

  list(token) {
    const ordered = token.ordered;
    const start = token.start;

    const out = document.createElement(ordered ? 'ol' : 'ul');
    for (let j = 0; j < token.items.length; j++) {
      const item = token.items[j];
      out.append(this.listitem(item));
    }

    if (ordered && start !== 1) {
      out.setAttribute('start', start);
    }

    return out;
  },

  listitem(item) {
    const out = document.createElement('li');
    out.append(this.parser.parse(item.tokens, !!item.loose));
    if (item.task) {
      const checkbox = this.checkbox({ checked: !!item.checked });
      if (item.loose) {
        if (item.tokens[0]?.type === 'text') {
          out.firstChild.prepend(document.createTextNode(' '));
          out.firstChild.prepend(checkbox);
        } else {
          out.prepend(document.createTextNode(' '));
          out.prepend(checkbox);
        }
      } else {
        out.prepend(document.createTextNode(' '));
        out.prepend(checkbox);
      }
    }

    return out;
  },

  checkbox({ checked }) {
    const checkbox = document.createElement('input');
    checkbox.type = 'checkbox';
    checkbox.checked = checked;
    if (checked) {
      checkbox.setAttribute('checked', '');
    }
    checkbox.disabled = true;
    return checkbox;
  },

  paragraph({ tokens }) {
    const paragraph = document.createElement('p');
    paragraph.append(this.parser.parseInline(tokens));
    return paragraph;
  },

  table(token) {
    const table = document.createElement('table');
    const thead = document.createElement('thead');

    const headerCell = document.createDocumentFragment();
    for (let j = 0; j < token.header.length; j++) {
      headerCell.append(this.tablecell(token.header[j]));
    }
    thead.append(this.tablerow({ text: headerCell }));
    table.append(thead);

    if (token.rows.length === 0) {
      return table;
    }

    const tbody = document.createElement('tbody');
    for (let j = 0; j < token.rows.length; j++) {
      const row = token.rows[j];

      const cell = document.createDocumentFragment();
      for (let k = 0; k < row.length; k++) {
        cell.append(this.tablecell(row[k]));
      }

      tbody.append(this.tablerow({ text: cell }));
    }

    table.append(tbody);

    return table;
  },

  tablerow({ text }) {
    const tr = document.createElement('tr');
    tr.append(text);
    return tr;
  },

  tablecell(token) {
    const content = this.parser.parseInline(token.tokens);
    const out = document.createElement(token.header ? 'th' : 'td');
    out.append(content);
    if (token.align) {
      out.setAttribute('align', token.align);
    }

    return out;
  },

  /**
   * span level renderer
   */
  strong({ tokens }) {
    const strong = document.createElement('strong');
    strong.append(this.parser.parseInline(tokens));
    return strong;
  },

  em({ tokens }) {
    const em = document.createElement('em');
    em.append(this.parser.parseInline(tokens));
    return em;
  },

  codespan({ text }) {
    const code = document.createElement('code');
    code.innerHTML = escapeText(text, true);
    return code;
  },

  br() {
    return document.createElement('br');
  },

  del({ tokens }) {
    const del = document.createElement('del');
    del.append(this.parser.parseInline(tokens));
    return del;
  },

  link({ href, title, tokens }) {
    const body = this.parser.parseInline(tokens);
    const cleanHref = cleanUrl(href);
    /* istanbul ignore next */
    if (cleanHref === null) {
      return body;
    }
    href = cleanHref;
    const out = document.createElement('a');
    out.href = href;
    if (title) {
      out.title = title;
    }
    out.append(body);
    return out;
  },

  image({ href, title, text, tokens }) {
    const body = this.parser.parseInline(tokens, textRenderer);

    const cleanHref = cleanUrl(href);
    /* istanbul ignore next */
    if (cleanHref === null) {
      return document.createTextNode(escapeText(body.textContent));
    }
    href = cleanHref;
    const out = document.createElement('img');
    out.src = href;
    out.alt = body.textContent;
    if (title) {
      out.title = escapeText(title);
    }
    out.append(body);
    return out;
  },

  text(token) {
    return token.tokens
      ? this.parser.parseInline(token.tokens)
      : document.createTextNode(token.text);
  },
};
