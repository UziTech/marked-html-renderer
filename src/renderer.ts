import type { Renderer, MarkedOptions, Parser } from 'marked';

export const other = {
  escapeTest: /[&<>"']/,
  escapeReplace: /[&<>"']/g,
  escapeTestNoEncode: /[<>"']|&(?!(#\d{1,7}|#[Xx][a-fA-F0-9]{1,6}|\w+);)/,
  escapeReplaceNoEncode: /[<>"']|&(?!(#\d{1,7}|#[Xx][a-fA-F0-9]{1,6}|\w+);)/g,
  percentDecode: /%25/g,
  notSpaceStart: /^\S*/,
  endingNewline: /\n$/,
};

const escapeReplacements: Record<string, string> = {
  '&': '&amp;',
  '<': '&lt;',
  '>': '&gt;',
  '"': '&quot;',
  "'": '&#39;',
} as const;
const getEscapeReplacement = (ch: string) => escapeReplacements[ch];

export function escapeText(html: string, encode?: boolean) {
  if (encode) {
    if (other.escapeTest.test(html)) {
      return html.replace(other.escapeReplace, getEscapeReplacement);
    }
  } else {
    if (other.escapeTestNoEncode.test(html)) {
      return html.replace(other.escapeReplaceNoEncode, getEscapeReplacement);
    }
  }

  return html;
}

export function cleanUrl(href: string) {
  return encodeURI(href).replace(other.percentDecode, '%');
}

export const renderer: Renderer<DocumentFragment, Node | string> = {
  options: null as unknown as MarkedOptions<DocumentFragment, Node | string>,
  parser: null as unknown as Parser<DocumentFragment, Node | string>,

  space() {
    return '';
  },

  code({ text, lang, escaped }) {
    const langString = (lang || '').match(other.notSpaceStart)?.[0];

    let code = text.replace(other.endingNewline, '') + '\n';
    code = (escaped ? code : escapeText(code, true));

    const preEl = document.createElement('pre');
    const codeEl = document.createElement('code');
    preEl.appendChild(codeEl);
    codeEl.textContent = code;

    if (langString) {
      preEl.classList.add('language-' + langString);
    }

    return preEl;
  },

  blockquote({ tokens }) {
    const blockquote = document.createElement('blockquote');
    blockquote.append(this.parser.parse(tokens));
    return blockquote;
  },

  html({ text }) {
    // HTML should be handled by the blockHtml and inlineHtml extensions in extension.js
    // Handle comments
    const comment = /^<!--([\s\S]*?)-->/.exec(text);
    if (comment) {
      return document.createComment(comment[1]);
    }
    // If it is not just assume it is text.
    return text;
  },

  heading({ tokens, depth }) {
    const heading = document.createElement('h' + depth);
    heading.append(this.parser.parseInline(tokens));
    return heading;
  },

  hr() {
    return document.createElement('hr');
  },

  list(token) {
    const ordered = token.ordered;
    const start = token.start.toString();

    const out = document.createElement(ordered ? 'ol' : 'ul');
    for (let j = 0; j < token.items.length; j++) {
      const item = token.items[j];
      out.append(this.listitem(item));
    }

    if (ordered && start !== '1') {
      out.setAttribute('start', start);
    }

    return out;
  },

  listitem(item) {
    const out = document.createElement('li');
    out.append(this.parser.parse(item.tokens, !!item.loose));
    if (item.task) {
      const checkbox = this.checkbox({ checked: !!item.checked });
      if (item.loose) {
        if (item.tokens[0]?.type === 'text') {
          (out.firstChild as HTMLElement).prepend(' ');
          (out.firstChild as HTMLElement).prepend(checkbox);
        } else {
          out.prepend(' ');
          out.prepend(checkbox);
        }
      } else {
        out.prepend(' ');
        out.prepend(checkbox);
      }
    }

    return out;
  },

  checkbox({ checked }) {
    const checkbox = document.createElement('input');
    checkbox.type = 'checkbox';
    checkbox.checked = checked;
    if (checked) {
      checkbox.setAttribute('checked', '');
    }
    checkbox.disabled = true;
    return checkbox;
  },

  paragraph({ tokens }) {
    const paragraph = document.createElement('p');
    paragraph.append(this.parser.parseInline(tokens));
    return paragraph;
  },

  table(token) {
    const table = document.createElement('table');
    const thead = document.createElement('thead');

    const headerCell = document.createDocumentFragment();
    for (let j = 0; j < token.header.length; j++) {
      headerCell.append(this.tablecell(token.header[j]));
    }
    thead.append(this.tablerow({ text: headerCell }));
    table.append(thead);

    if (token.rows.length === 0) {
      return table;
    }

    const tbody = document.createElement('tbody');
    for (let j = 0; j < token.rows.length; j++) {
      const row = token.rows[j];

      const cell = document.createDocumentFragment();
      for (let k = 0; k < row.length; k++) {
        cell.append(this.tablecell(row[k]));
      }

      tbody.append(this.tablerow({ text: cell }));
    }

    table.append(tbody);

    return table;
  },

  tablerow({ text }) {
    const tr = document.createElement('tr');
    tr.append(text);
    return tr;
  },

  tablecell(token) {
    const content = this.parser.parseInline(token.tokens);
    const out = document.createElement(token.header ? 'th' : 'td');
    out.append(content);
    if (token.align) {
      out.setAttribute('align', token.align);
    }

    return out;
  },

  /**
   * span level renderer
   */
  strong({ tokens }) {
    const strong = document.createElement('strong');
    strong.append(this.parser.parseInline(tokens));
    return strong;
  },

  em({ tokens }) {
    const em = document.createElement('em');
    em.append(this.parser.parseInline(tokens));
    return em;
  },

  codespan({ text }) {
    const code = document.createElement('code');
    code.innerHTML = escapeText(text, true);
    return code;
  },

  br() {
    return document.createElement('br');
  },

  del({ tokens }) {
    const del = document.createElement('del');
    del.append(this.parser.parseInline(tokens));
    return del;
  },

  link({ href, title, tokens }) {
    const body = this.parser.parseInline(tokens);
    href = cleanUrl(href);
    const out = document.createElement('a');
    out.href = href;
    if (title) {
      out.title = title;
    }
    out.append(body);
    return out;
  },

  image({ href, title, text, tokens }) {
    const body = this.parser.parseInline(tokens, this.parser.textRenderer);

    href = cleanUrl(href);
    const out = document.createElement('img');
    out.src = href;
    out.alt = body.textContent ?? '';
    if (title) {
      out.title = escapeText(title);
    }
    out.append(body);
    return out;
  },

  text(token) {
    return ('tokens' in token) && token.tokens
      ? this.parser.parseInline(token.tokens)
      : document.createTextNode(token.text);
  },
};
