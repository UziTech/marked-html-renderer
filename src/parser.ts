import type { MarkedOptions, Token, Tokens, MarkedToken, TextRenderer, Renderer, Parser as MarkedParser } from 'marked';
import { textRenderer } from './textRenderer.ts';
import { renderer } from './renderer.ts';

export class Parser {
  options: MarkedOptions<DocumentFragment, Node | string>;
  renderer: Renderer<DocumentFragment, Node | string>;
  textRenderer: TextRenderer<Node | string>;

  constructor(options: MarkedOptions<DocumentFragment, Node | string> | undefined) {
    this.options = options ?? { renderer };
    this.textRenderer = textRenderer;
    this.renderer = this.options.renderer!;
    this.renderer.parser = this as unknown as MarkedParser<DocumentFragment, Node | string>;
  }

  /**
   * Static Parse Method
   */
  static parse(tokens: Token[], options?: MarkedOptions<DocumentFragment, Node | string>) {
    const parser = new Parser(options);
    return parser.parse(tokens);
  }

  /**
   * Static Parse Inline Method
   */
  static parseInline(tokens: Token[], options?: MarkedOptions<DocumentFragment, Node | string>) {
    const parser = new Parser(options);
    return parser.parseInline(tokens);
  }

  appendOutput(out: DocumentFragment, ret: Node | string | false | undefined) {
    if (ret) {
      out.append(ret);
    }
  }

  parse(tokens: Token[], top = true) {
    const out = document.createDocumentFragment();

    for (let i = 0; i < tokens.length; i++) {
      const anyToken = tokens[i];

      // Run any renderer extensions
      if (this.options.extensions?.renderers?.[anyToken.type]) {
        const genericToken = anyToken;
        const ret = this.options.extensions.renderers[genericToken.type].call({ parser: this as unknown as MarkedParser<DocumentFragment, Node | string> }, genericToken);
        if (ret !== false || !['space', 'hr', 'heading', 'code', 'table', 'blockquote', 'list', 'html', 'paragraph', 'text'].includes(genericToken.type)) {
          this.appendOutput(out, ret);
          continue;
        }
      }

      const token = anyToken as MarkedToken;

      switch (token.type) {
        case 'space': {
          this.appendOutput(out, this.renderer!.space(token));
          continue;
        }
        case 'hr': {
          this.appendOutput(out, this.renderer!.hr(token));
          continue;
        }
        case 'heading': {
          this.appendOutput(out, this.renderer!.heading(token));
          continue;
        }
        case 'code': {
          this.appendOutput(out, this.renderer!.code(token));
          continue;
        }
        case 'table': {
          this.appendOutput(out, this.renderer!.table(token));
          continue;
        }
        case 'blockquote': {
          this.appendOutput(out, this.renderer!.blockquote(token));
          continue;
        }
        case 'list': {
          this.appendOutput(out, this.renderer!.list(token));
          continue;
        }
        case 'html': {
          this.appendOutput(out, this.renderer!.html(token));
          continue;
        }
        case 'def': {
          this.appendOutput(out, this.renderer!.def(token));
          continue;
        }
        case 'paragraph': {
          this.appendOutput(out, this.renderer!.paragraph(token));
          continue;
        }
        case 'text': {
          let textToken = token;
          const body = document.createDocumentFragment();

          body.append(this.renderer!.text(textToken));
          while (i + 1 < tokens.length && tokens[i + 1].type === 'text') {
            textToken = tokens[++i] as Tokens.Text;
            body.append('\n', this.renderer!.text(textToken));
          }
          if (top) {
            const p = this.renderer!.paragraph({
              type: 'paragraph',
              raw: '',
              text: '',
              tokens: [],
            }) as unknown as HTMLParagraphElement;
            p.append(body);
            this.appendOutput(out, p);
          } else {
            this.appendOutput(out, body);
          }
          continue;
        }

        default: {
          const errMsg = 'Token with "' + token.type + '" type was not found.';
          if (this.options.silent) {
            console.error(errMsg);
            return document.createDocumentFragment();
          } else {
            throw new Error(errMsg);
          }
        }
      }
    }

    return out;
  }

  /**
   * Parse Inline Tokens
   */
  parseInline(tokens: Token[], renderer = this.renderer) {
    const out = document.createDocumentFragment();

    for (let i = 0; i < tokens.length; i++) {
      const anyToken = tokens[i];

      // Run any renderer extensions
      if (this.options.extensions?.renderers?.[anyToken.type]) {
        const ret = this.options.extensions.renderers[anyToken.type].call({ parser: this as unknown as MarkedParser<DocumentFragment, Node | string> }, anyToken);
        if (ret !== false || !['escape', 'html', 'link', 'image', 'strong', 'em', 'codespan', 'br', 'del', 'text'].includes(anyToken.type)) {
          this.appendOutput(out, ret);
          continue;
        }
      }

      const token = anyToken as MarkedToken;

      switch (token.type) {
        case 'escape': {
          this.appendOutput(out, renderer!.text(token));
          break;
        }
        case 'html': {
          this.appendOutput(out, renderer!.html(token));
          break;
        }
        case 'link': {
          this.appendOutput(out, renderer!.link(token));
          break;
        }
        case 'image': {
          this.appendOutput(out, renderer!.image(token));
          break;
        }
        case 'strong': {
          this.appendOutput(out, renderer!.strong(token));
          break;
        }
        case 'em': {
          this.appendOutput(out, renderer!.em(token));
          break;
        }
        case 'codespan': {
          this.appendOutput(out, renderer!.codespan(token));
          break;
        }
        case 'br': {
          this.appendOutput(out, renderer!.br(token));
          break;
        }
        case 'del': {
          this.appendOutput(out, renderer!.del(token));
          break;
        }
        case 'text': {
          this.appendOutput(out, renderer!.text(token));
          break;
        }
        default: {
          const errMsg = 'Token with "' + token.type + '" type was not found.';
          if (this.options.silent) {
            console.error(errMsg);
            return document.createDocumentFragment();
          } else {
            throw new Error(errMsg);
          }
        }
      }
    }
    return out;
  }
}
