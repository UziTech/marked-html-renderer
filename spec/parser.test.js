import { Parser } from '../src/parser.js';
import { renderer } from '../src/renderer.js';
import { getInnerHTML } from './helpers.js';

describe('Parser', () => {
  test('multiple text tokens', () => {
    const parser = new Parser({ renderer });
    const tokens = [
      {
        type: 'text',
        raw: 'text',
        text: 'text',
      },
      {
        type: 'text',
        raw: 'text2',
        text: 'text2',
      },
    ];

    expect(getInnerHTML(parser.parse(tokens))).toMatchSnapshot();
  });

  test('escaped code', () => {
    const parser = new Parser({ renderer });
    const tokens = [
      {
        type: 'code',
        raw: 'code',
        text: 'code',
        escaped: true,
      },
    ];

    expect(getInnerHTML(parser.parse(tokens))).toMatchSnapshot();
  });

  test('invalid block token type', () => {
    const parser = new Parser({ renderer });
    const tokens = [
      {
        type: 'invalid',
      },
    ];

    expect(() => getInnerHTML(parser.parse(tokens))).toThrow('Token with "invalid" type was not found.');
  });

  test('invalid block token type silent', () => {
    const parser = new Parser({ renderer, silent: true });
    const tokens = [
      {
        type: 'invalid',
      },
    ];

    expect(getInnerHTML(parser.parse(tokens))).toMatchSnapshot();
  });

  test('invalid inline token type', () => {
    const parser = new Parser({ renderer });
    const tokens = [
      {
        type: 'invalid',
      },
    ];

    expect(() => getInnerHTML(parser.parseInline(tokens))).toThrow('Token with "invalid" type was not found.');
  });

  test('invalid inline token type silent', () => {
    const parser = new Parser({ renderer, silent: true });
    const tokens = [
      {
        type: 'invalid',
      },
    ];

    expect(getInnerHTML(parser.parseInline(tokens))).toMatchSnapshot();
  });
});
