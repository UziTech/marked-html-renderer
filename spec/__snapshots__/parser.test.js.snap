// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Parser escaped code 1`] = `
"<pre><code>code
</code></pre>"
`;

exports[`Parser invalid block token type silent 1`] = `""`;

exports[`Parser invalid inline token type silent 1`] = `""`;

exports[`Parser multiple text tokens 1`] = `
"<p>text
text2</p>"
`;
